dataflow/src/main/java/org/checkerframework/dataflow/busyexpression/BusyExprValue.java

```diff
@@ -0,0 +1,49 @@
+package org.checkerframework.dataflow.busyexpression;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.checkerframework.dataflow.analysis.AbstractValue;
+import org.checkerframework.dataflow.cfg.node.Node;
+import org.checkerframework.javacutil.BugInCF;
+
+/** A busy expression wrapper turning a node into the abstract value */
+public class BusyExprValue implements AbstractValue<BusyExprValue> {
+
+    /**
+     * A busy expression is represented by a node, which can be a {@link
+     * org.checkerframework.dataflow.cfg.node.BinaryOperationNode}
+     */
+    protected final Node busyExpression;
+
+    @Override
+    public BusyExprValue leastUpperBound(BusyExprValue other) {
+        throw new BugInCF("lub of BusyExpValue get called!");
+    }
+
+    /**
+     * Create a new busy expression.
+     *
+     * @param n a node
+     */
+    public BusyExprValue(Node n) {
+        this.busyExpression = n;
+    }
+
+    @Override
+    public String toString() {
+        return this.busyExpression.toString();
+    }
+
+    @Override
+    public int hashCode() {
+        return this.busyExpression.hashCode();
+    }
+
+    @Override
+    public boolean equals(@Nullable Object obj) {
+        if (!(obj instanceof BusyExprValue)) {
+            return false;
+        }
+        BusyExprValue other = (BusyExprValue) obj;
+        return this.busyExpression.equals(other.busyExpression);
```

Comment:
Nodes should not be compared with `.equals`, only with `==`. This probably isn't a problem here, but in general, two nodes can contain the same values, but their identity is different and they should be treated differently.

Reply:
Got it, thanks for pointing this out!

I will fix the problem in liveVariable analysis too.

Reply:
Sorry, after changing this, the testcase failed. The `equals` function here will be used in `leastUpperBound` function in the `BusyExprStore.java`. If we treat differently for two nodes with the same value but with different identity, the least upper bound function will return null when intersecting two stores.

Reply:
`.equals` checks the value equality rather than reference equality.
Consider the scenario, we have expression `a-b` used in both the if and else branch, but one with field variable `a` and another with local variable `a`. Will the program mistakenly take two different `a-b` as equal and produce this as the busy expression?
Some corner cases checked:
Two local variables with the same name declared and used in if and else branch.
Local variable shadows a field variable

Reply:
It looks like using `equals` is faulty. Suppose we have 2 expressions
`b + foo()` both in the `if` and `else` branches, where calling `foo()`
returns different values each time. It seems that our program will analyze
the 2 expressions as the same and make it as the busy expression at the
conditional block. However, foo() is a method invocation, and calling
it will introduce an exceptional block follows it. As the result,
there are at least 2 blocks following foo() block, one with a transfer
result `b + foo()`, another is the exception block with empty transfer
result. Their intersection is empty when propagating to the foo() block.
The abstract value `b+foo()` is lost during backward propagating, hence
it's not a problem in this case.